### PR TITLE
Ticket 23521 Support changing model bases in migrations

### PR DESCRIPTION
**Pull request by Zorking**
Not sure how can I edit old and closed PR from another user so I created a new one.

Original PR is here: https://github.com/django/django/pull/11222
Ticket: https://code.djangoproject.com/ticket/23521

I'm actually using the code in my project, decided to fix it here.

### CODE REVIEW

**Comment by felixxm** [comment]
Thanks for this patch :+1: Simon's [comment](https://github.com/django/django/pull/11222#pullrequestreview-233821387) is not addressed.

**Comment by Zorking** [comment]
@felixxm created autodetect for changing bases but found another issue with model bases, which is kinda blocking some of `AlterModelBases` functionality. I can't find a ticket for that, should I address it in that PR? 

**Steps to reproduce:**
1) Create 2 models that have models.Model in bases:
class A(models.Model):
    pass
class B(models.Model):
    pass

2) Try to apply multi-table inheritance to one of them:
 class B(A):
    pass

**Results:**
You will get exception:
`django.core.exceptions.FieldError: Local field 'id' in class 'B' clashes with field of the same name from base class 'A'.`
 btw, if you have more than one model in bases, you will get the similar error also

I can see 2 possible solutions: 
1) Rename ID field, change it to models.OneToOne with primary_key and make sure the user will prepopulate that field somehow.
2) Just raise the exception and ask the user to write migration manually.

**Comment by jacobtylerwalls** [comment]
hiya @Zorking and happy new year. Distinct issue deserve distinct tickets! 👍🏻  This allows folks who monitor tickets a chance to vet your issue and accept it or ask for clarification. If your issue is accepted you can claim it and fix here; or if your fix is lightweight you might choose to fix in a separate PR that can get merged fast. Best, Jacob

**Comment by Zorking** [comment]
> hiya @Zorking and happy new year. Distinct issue deserve distinct tickets! 👍🏻 This allows folks who monitor tickets a chance to vet your issue and accept it or ask for clarification. If your issue is accepted you can claim it and fix here; or if your fix is lightweight you might choose to fix in a separate PR that can get merged fast. Best, Jacob

Hi!

The issue was created 6 years ago and [the ticket](https://code.djangoproject.com/ticket/23521) is in detail of the PR. 

I already tested the fix in the production of the project that I worked on, worked great. 

Should I change the PR name to something different or add some tags?

**Comment by jacobtylerwalls** [comment]
Ah, I was referring to the comment directly above where you said you identified a second issue without a ticket.

**Comment by Zorking** [comment]
@jacobtylerwalls got it!

Is it worth spending time with? Maybe I didn't get the workflow right, but the PR and the ticket look kinda dead.

**Comment by jacobtylerwalls** [comment]
Short answer is yes. Long answer is that some new Django contributors have some trouble getting used to the setting Needs... flags on the ticket to No to get in the patch review queue. I can recommend you set those flags once we get this other issue to no longer be blocking! Cheers.

**Comment by Zorking** [comment]
@jacobtylerwalls I created [the ticket](https://code.djangoproject.com/ticket/32330). 

> once we get this other issue to no longer be blocking! 

The issue is not blocking the current fix, it's just limiting it a bit. `AlterModelBases` will remove model inheritance, but can't be used in order to actually alter all possible model bases (add new, patch, etc.). That is what I meant by blocking some of the functionality.

**Comment by jacobtylerwalls** [comment]
Super. Then if in your judgment this patch is ready for review please set all the Needs... flags in the ticket to no.

File: tests/migrations/test_operations.py [hunk 1]
```diff
@@ -1278,6 +1278,25 @@ def test_alter_model_table_m2m(self):
...
+        """
+        project_state = self.set_up_test_model("test_almotb", mti_model=True)
+        # Test the state alteration
+        operation = migrations.AlterModelBases("shetlandpony", bases=(models.Model,))
+        self.assertEqual(operation.describe(), "Update shetlandpony bases to {}".format((models.Model,)))
+        new_state = project_state.clone()
+        operation.state_forwards("test_almotb", new_state)
+        self.assertEqual(new_state.models["test_almotb", "shetlandpony"].bases, (models.Model,))
+        # Test the database alteration
+        # And test reversal
```
Comment (felixxm) [inline tests/migrations/test_operations.py:1293]: Tests are missing.

File: docs/ref/migration-operations.txt [hunk 2]
```diff
@@ -132,6 +132,18 @@ should be a dictionary mapping option names to values.
 
 Alters the managers that are available during migrations.
 
+``AlterModelBases``
+-------------------
+
+.. class:: AlterModelBases(model_name, bases)
+
+.. versionadded:: 3.0
```
Comment (felixxm) [inline docs/ref/migration-operations.txt:140]: Please address it to the Django 4.0.

**Review by felixxm — COMMENTED** [review]
@Zorking Thanks for updates :+1: Can you check duplicates of this ticket (ticket-31343, ticket-31329, ticket-30513, ticket-26488) and add tests (if not covered) for described scenarios

**Comment by felixxm** [comment]
Closing due to inactivity.

**Comment by Pomax** [comment]
What? No, please reopen this, this is an _insanely_ important thing to fix =(

**Comment by felixxm** [comment]
@Pomax Feel-free to check comment and prepare a new patch.

**Comment by Pomax** [comment]
If I had the skill, I would, but this is exactly the thing I _don't_ have the skills for to the point where I had to bounty off 500 rep on Stackoverflow just to get help even getting a workaround. Having the team take this one on instead of external contributor at this point in the PR is most definitely worth the team's time given what it fixes (and how long that bug's been the case).

**Comment by charettes** [comment]
> Having the team take this one on instead of relying on external contributor at this point in the PR is most definitely worth the team's time given what it fixes

The _team_ **is** external contributors and your feeling about the urgency of fixing this particular issue is shared amongst all Django users affected a particular issue affecting their workflow. In other words nothing makes it issue _special_ compared to [the other existing hundreds of bugs](https://dashboard.djangoproject.com/) and the most productive way of getting it fixed it to commit time to making the few remaining adjustments brought up above as their is no paid contributors actively fixing bugs that aren't issues affecting Django releases (regressions, security issues)

Closing inactive PRs is just a way for fellows to manage the large number of changes requests received by the community in a way that ensures the sustainability of the project in terms of code quality and maintainability. To give you an example, if a developer proposing non-trivial changes is not responsive during the review process it's likely that they also won't be if their changes happen to introduce a regression that the limited number of volunteers and the fellows will have to figure out a solution for or simply revert the patch.

**Comment by Pomax** [comment]
While true, what does make it special is the fact that there already is a PR that's 95% of the way there, based on the comments? As opposed to bugs that haven't been started on. @Zorking are you still on github at all? If so, could I convince you to finish this PR?

**Pull request closed without merging**